File: 2021/d3/python/d3p2_solution.py
# Advent of Code - Day 3 Problem 2
# Extremely brute-force solution. May work on something more clean (possibly with recursion)

inputFile = open ('../input/d3_input.txt','r')

listOfBinary = inputFile.readlines()
listOfBinary2 = listOfBinary.copy()

lengthofBinary = len(list(listOfBinary[0])) - 1 # Sub one to account for appended /n
emptylist = [None] * lengthofBinary

# Function which takes in a list making up binary number and converts it into an int in decimal form
def convertToDecimal(binaryNumList):
    decimal = 0
    for i in range(lengthofBinary):
        decimal += (int(binaryNumList[i]) * (2 ** (lengthofBinary -1 - i)))
    return decimal
    
# Function to remove binary number from list if it contains a zero in the indicated index
def removeZeros(listOfBinary, index):
    for i in range(len(listOfBinary)):
        if list(listOfBinary[i])[index] == "0":
            del listOfBinary[i]
            listOfBinary.insert(0,emptylist) # Adding empty item equal in size to not disrupt list/array size
    return listOfBinary

# Function to remove binary number from list if it contains a one in the indicated index
def removeOnes(listOfBinary, index):
    for i in range(len(listOfBinary)):
        if list(listOfBinary[i])[index] == "1":
            del listOfBinary[i]
            listOfBinary.insert(0,emptylist) # Adding empty item equal in size to not disrupt list/array size
    return listOfBinary


# Calculate oxygen generator rating by iterating through list of binary numbers and removing items from the list 
# based on the most occuring number within each index
for i in range(lengthofBinary):
    numzeros=0
    numones=0
    for j in range(len(listOfBinary)): 
            binary = list(listOfBinary[j])
            if binary[i] == "1":
                numones+=1
            if binary[i] == "0":
                numzeros += 1 
    
    # Break early if there is only one number left
    if listOfBinary[-2] == emptylist: 
            break

    # Check cases for number of zeros & ones in order to determine what binary number must be removed from list
    if numones > numzeros:
        listOfBinary = removeZeros(listOfBinary, i)
    elif numones < numzeros:
        listOfBinary = removeOnes(listOfBinary, i)
    elif numones == numzeros:
        listOfBinary = removeZeros(listOfBinary, i)

oxygenGenerator = convertToDecimal(list(listOfBinary[-1]))

# Calculate CO2 scrubber rating by iterating through list of binary numbers and removing items from the list 
# based on the least occuring number within each index
for i in range(lengthofBinary):
    numzeros=0
    numones=0
    for j in range(len(listOfBinary2)):
        binary = list(listOfBinary2[j])
        if binary[i] == "1":
            numones+=1
        if binary[i] == "0":
            numzeros += 1 
    
    # Break early if there is only one number left
    if listOfBinary2[-2] == emptylist: 
            break
    
    # Check cases for number of zeros & ones in order to determine what binary number must be removed from list
    if numones > numzeros:
        listOfBinary2 = removeOnes(listOfBinary2, i)
    elif numones < numzeros:
        listOfBinary2 = removeZeros(listOfBinary2, i)
    elif numones == numzeros:
        listOfBinary2 = removeOnes(listOfBinary2, i)

carbonScrubber = convertToDecimal(list(listOfBinary2[-1]))

# Calculate life support rating and print results
print(oxygenGenerator * carbonScrubber)

inputFile.close() 

File: 2021/d3/cpp/d3p2_solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

// Opens file and converts file input to vector
void getFileInput(vector<string> &fileInput)
{
    string line;

    fstream file("../input/d3_input.txt");

    if (!file)
        cout << "Error opening file!";
    else
        while (getline(file, line))
            fileInput.push_back(line);
}

void removeOnes(vector<string> &binaryList, int index)
{
    for (int i=0; i < binaryList.size(); i++)
    {
        if (binaryList.at(i)[index] == '1')
        {
            binaryList.erase(binaryList.begin() + i);
            i--;
        }
    }
}

void removeZeros(vector<string> &binaryList, int index)
{
     for (int i=0; i < binaryList.size(); i++)
    {
        if (binaryList.at(i)[index] == '0')
        {
            binaryList.erase(binaryList.begin() + i);
            i--;
        }
    }
}

void calculateCO2 (vector<string> &co2Scrub)
{
    int numOnes=0, numZeros=0;

    for(int i = 0; i < co2Scrub.at(0).size(); i++)
    {
        for(int j=0; j < co2Scrub.size(); j++)
        {
            if (co2Scrub.at(j)[i] == '0')
                numZeros++;
            else if (co2Scrub.at(j)[i] == '1')
                numOnes++;
        }
        
        if (co2Scrub.size() == 1)
            return;
        else if (numZeros > numOnes)
            removeZeros(co2Scrub, i);

        else
            removeOnes(co2Scrub, i);

        numZeros=0;
        numOnes=0;
    }
}

void calculateOxygen(vector<string> &oxygenGenRating)
{
    int numOnes=0, numZeros=0;

    for(int i = 0; i < oxygenGenRating.at(0).size(); i++)
    {
        for(int j=0; j < oxygenGenRating.size(); j++)
        {
            if (oxygenGenRating.at(j)[i] == '0')
                numZeros++;
            else if (oxygenGenRating.at(j)[i] == '1')
                numOnes++;
        }
        
        if (oxygenGenRating.size() == 1)
            return;
        else if (numZeros > numOnes)
            removeOnes(oxygenGenRating, i);
        else
            removeZeros(oxygenGenRating, i);
        numZeros=0;
        numOnes=0;
    }
}

int main ()
{
    vector<string> oxygenGenRating;
    vector<string> co2Scrub;
    
    getFileInput(oxygenGenRating);

    co2Scrub = oxygenGenRating;
    
    calculateOxygen(oxygenGenRating);
    calculateCO2(co2Scrub);

    cout << "Oxygen Generator Rating: " << stoi(oxygenGenRating.at(0), 0, 2) << endl;
    cout << "CO2 Scrubber Rating: " << stoi(co2Scrub.at(0), 0, 2) << endl;
    cout << "Life Support: " << stoi(oxygenGenRating.at(0), 0, 2) * stoi(co2Scrub.at(0), 0, 2) << endl;

    return 0;
}